azurerm/resource_arm_sql_failover_group.go

```diff
@@ -0,0 +1,356 @@
+	client := meta.(*ArmClient).sqlFailoverGroupsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resourceGroup := id.ResourceGroup
+	serverName := id.Path["servers"]
+	name := id.Path["failoverGroups"]
+
+	future, err := client.Delete(ctx, resourceGroup, serverName, name)
+	if err != nil {
+		return fmt.Errorf("Error deleting SQL Failover Group %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
+	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		return fmt.Errorf("Error deleting SQL Failover Group %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
+	return err
+}
+
+func expandReadWritePolicy(d *schema.ResourceData) *sql.FailoverGroupReadWriteEndpoint {
+	vs := d.Get("read_write_endpoint_failover_policy").([]interface{})
+	v := vs[0].(map[string]interface{})
+
+	mode := sql.ReadWriteEndpointFailoverPolicy(v["mode"].(string))
+	graceMins := int32(v["grace_minutes"].(int))
+
+	policy := &sql.FailoverGroupReadWriteEndpoint{
+		FailoverPolicy: mode,
+	}
+
+	if mode != "manual" {
+		policy.FailoverWithDataLossGracePeriodMinutes = utils.Int32(graceMins)
+	}
+
+	return policy
+}
+
+func flattenReadWritePolicy(input *sql.FailoverGroupReadWriteEndpoint) []interface{} {
+	if input == nil {
+		return []interface{}{}
+	}
+
+	policy := make(map[string]interface{})
+
+	policy["mode"] = string(input.FailoverPolicy)
+
+	if input.FailoverWithDataLossGracePeriodMinutes != nil {
+		policy["grace_minutes"] = *input.FailoverWithDataLossGracePeriodMinutes
+	}
+	return []interface{}{policy}
+}
+
+func expandReadOnlyPolicy(d *schema.ResourceData) *sql.FailoverGroupReadOnlyEndpoint {
+	vs := d.Get("readonly_endpoint_failover_policy").([]interface{})
+	if len(vs) == 0 {
+		return nil
+	}
+
+	v := vs[0].(map[string]interface{})
+	mode := sql.ReadOnlyEndpointFailoverPolicy(v["mode"].(string))
+
+	policy := &sql.FailoverGroupReadOnlyEndpoint{
+		FailoverPolicy: mode,
+	}
+
+	return policy
+}
+
+func flattenReadOnlyPolicy(input *sql.FailoverGroupReadOnlyEndpoint) []interface{} {
+	if input == nil {
+		return []interface{}{}
+	}
+
+	policy := make(map[string]interface{})
+	policy["mode"] = string(input.FailoverPolicy)
+
+	return []interface{}{policy}
+}
+
+func expandPartnerServers(d *schema.ResourceData) *[]sql.PartnerInfo {
+	servers := d.Get("partner_servers").([]interface{})
+	partners := make([]sql.PartnerInfo, 0)
+
+	for _, server := range servers {
+		info := server.(map[string]interface{})
+
+		id := info["id"].(string)
+		partners = append(partners, sql.PartnerInfo{
+			ID: &id,
+		})
+	}
+
+	return &partners
+}
+
+func flattenPartnerServers(input *[]sql.PartnerInfo) []map[string]interface{} {
+	result := make([]map[string]interface{}, 0)
+
+	if input != nil {
+		for _, server := range *input {
+			info := make(map[string]interface{})
+			info["id"] = *server.ID
```

Comment:
We should be nil checking this
```suggestion
	if v :=server.ID; v != nil {
			info["id"] = *v
	}
```